diff --git a/common.gypi b/common.gypi
index adbd826a4f..963117e459 100644
--- a/common.gypi
+++ b/common.gypi
@@ -141,7 +141,7 @@
           ],
         },
         'defines': [ 'DEBUG', '_DEBUG', 'V8_ENABLE_CHECKS' ],
-        'cflags': [ '-g', '-O0' ],
+        'cflags': [ '-g', '-O0', '-fno-omit-frame-pointer' ],
         'conditions': [
           ['OS=="aix"', {
             'cflags': [ '-gxcoff' ],
@@ -487,6 +487,10 @@
             'defines': [ '_GLIBCXX_USE_C99_MATH' ],
             'libraries': [ '-llog' ],
           }],
+          ['_toolset=="target" and target_arch == "arm"', {
+            'defines': [ '_LIBCPP_HAS_NO_OFF_T_FUNCTIONS' ],
+          }],
+          
           ['_toolset=="host"', {
             'cflags': [ '-pthread' ],
             'ldflags': [ '-pthread' ],
