diff --git a/configure.py b/configure.py
index 47137d8912..d880227b44 100755
--- a/configure.py
+++ b/configure.py
@@ -1339,13 +1339,13 @@ def configure_node(o):
   if options.dest_os == 'android':
     shlib_suffix = 'so'
   elif sys.platform == 'darwin':
-    shlib_suffix = '%s.dylib'
+    shlib_suffix = 'dylib'
   elif sys.platform.startswith('aix'):
-    shlib_suffix = '%s.a'
+    shlib_suffix = 'a'
   elif sys.platform.startswith('zos'):
-    shlib_suffix = '%s.x'
+    shlib_suffix = 'x'
   else:
-    shlib_suffix = 'so.%s'
+    shlib_suffix = 'so'
   if '%s' in shlib_suffix:
     shlib_suffix %= node_module_version
 
