diff --git a/deps/uv/uv.gyp b/deps/uv/uv.gyp
index f03b4b4c0f..fd2bbb76b2 100644
--- a/deps/uv/uv.gyp
+++ b/deps/uv/uv.gyp
@@ -253,6 +253,7 @@
         }],
         [ 'OS=="android"', {
           'sources': [
+            'src/unix/epoll.c',
             'src/unix/linux-core.c',
             'src/unix/linux-inotify.c',
             'src/unix/linux-syscalls.c',
@@ -262,6 +263,7 @@
             'src/unix/procfs-exepath.c',
             'src/unix/random-getrandom.c',
             'src/unix/random-sysctl-linux.c',
+            'src/unix/random-getentropy.c',
           ],
           'link_settings': {
             'libraries': [ '-ldl' ],
