diff --git a/deps/openssl/openssl_asm.gypi b/deps/openssl/openssl_asm.gypi
index 2321fa1515..90f1f70a61 100644
--- a/deps/openssl/openssl_asm.gypi
+++ b/deps/openssl/openssl_asm.gypi
@@ -12,9 +12,9 @@
       'includes': ['config/archs/linux-ppc64/asm/openssl.gypi'],
     }, 'target_arch=="s390x" and OS=="linux"', {
       'includes': ['config/archs/linux64-s390x/asm/openssl.gypi'],
-    }, 'target_arch=="arm" and OS=="linux"', {
+    }, 'target_arch=="arm" and OS in "linux, android"', {
       'includes': ['config/archs/linux-armv4/asm/openssl.gypi'],
-    }, 'target_arch=="arm64" and OS=="linux"', {
+    }, 'target_arch=="arm64" and OS in "linux, android"', {
       'includes': ['config/archs/linux-aarch64/asm/openssl.gypi'],
     }, 'target_arch=="arm64" and OS=="mac"', {
       'includes': ['config/archs/darwin64-arm64-cc/asm/openssl.gypi'],
