diff --git a/deps/uv/uv.gyp b/deps/uv/uv.gyp
index 47de0d7ede..5fd1b40a00 100644
--- a/deps/uv/uv.gyp
+++ b/deps/uv/uv.gyp
@@ -254,6 +254,7 @@
         }],
         [ 'OS=="android"', {
           'sources': [
+            'src/unix/epoll.c',
             'src/unix/linux-core.c',
             'src/unix/linux-inotify.c',
             'src/unix/linux-syscalls.c',
@@ -263,6 +264,7 @@
             'src/unix/procfs-exepath.c',
             'src/unix/random-getrandom.c',
             'src/unix/random-sysctl-linux.c',
+            'src/unix/random-getentropy.c',
           ],
           'link_settings': {
             'libraries': [ '-ldl' ],
