diff --git a/deps/openssl/config/archs/linux-aarch64/no-asm/openssl-cl.gypi b/deps/openssl/config/archs/linux-aarch64/no-asm/openssl-cl.gypi
index 50b7d77b0c..8c960c90df 100644
--- a/deps/openssl/config/archs/linux-aarch64/no-asm/openssl-cl.gypi
+++ b/deps/openssl/config/archs/linux-aarch64/no-asm/openssl-cl.gypi
@@ -71,9 +71,13 @@
   },
   'defines': ['<@(openssl_defines_linux-aarch64)'],
   'include_dirs': [
+    '.',
     './include',
+    './crypto',
+    './crypto/include/internal',
   ],
   'cflags' : ['<@(openssl_cflags_linux-aarch64)'],
   'libraries': ['<@(openssl_ex_libs_linux-aarch64)'],
-  'sources': ['<@(openssl_cli_srcs_linux-aarch64)'],
+  'includes': ['./openssl.gypi'],
+  'sources': ['<@(openssl_sources)', '<@(openssl_sources_linux-aarch64)', '<@(openssl_cli_srcs_linux-aarch64)'],
 }
