diff --git a/deps/v8/src/api/api.cc b/deps/v8/src/api/api.cc
index 031d4244f7..f40604850a 100644
--- a/deps/v8/src/api/api.cc
+++ b/deps/v8/src/api/api.cc
@@ -5885,7 +5885,7 @@ bool TryHandleWebAssemblyTrapPosix(int sig_code, siginfo_t* info,
   // code rather than the wasm code, so the trap handler cannot find the landing
   // pad and lets the process crash. Therefore, only enable trap handlers if
   // the host and target arch are the same.
-#if (V8_TARGET_ARCH_X64 && !V8_OS_ANDROID) || \
+#if (V8_TARGET_ARCH_X64 && !V8_TARGET_OS_ANDROID) || \
     (V8_HOST_ARCH_ARM64 && V8_TARGET_ARCH_ARM64 && V8_OS_MACOSX)
   return i::trap_handler::TryHandleSignal(sig_code, info, context);
 #else
