diff --git a/deps/openssl/openssl_asm_avx2.gypi b/deps/openssl/openssl_asm_avx2.gypi
index 57a5c5c3c6..eb089af761 100644
--- a/deps/openssl/openssl_asm_avx2.gypi
+++ b/deps/openssl/openssl_asm_avx2.gypi
@@ -12,9 +12,9 @@
       'includes': ['config/archs/linux-ppc64/asm_avx2/openssl.gypi'],
     }, 'target_arch=="s390x" and OS=="linux"', {
       'includes': ['config/archs/linux64-s390x/asm_avx2/openssl.gypi'],
-    }, 'target_arch=="arm" and OS=="linux"', {
+    }, 'target_arch=="arm" and OS in "linux, android"', {
       'includes': ['config/archs/linux-armv4/asm_avx2/openssl.gypi'],
-    }, 'target_arch=="arm64" and OS=="linux"', {
+    }, 'target_arch=="arm64" and OS in "linux, android"', {
       'includes': ['config/archs/linux-aarch64/asm_avx2/openssl.gypi'],
     }, 'target_arch=="ia32" and OS=="freebsd"', {
       'includes': ['config/archs/BSD-x86/asm_avx2/openssl.gypi'],
