diff --git a/src/env-inl.h b/src/env-inl.h
index e679780900..a5e91eba2e 100644
--- a/src/env-inl.h
+++ b/src/env-inl.h
@@ -1233,6 +1233,7 @@ void Environment::set_main_utf16(std::unique_ptr<v8::String::Value> str) {
 
 void Environment::set_process_exit_handler(
     std::function<void(Environment*, int)>&& handler) {
+  fprintf(stderr, "set_process_exit_handler: %p\n", &handler);
   process_exit_handler_ = std::move(handler);
 }
 
