diff --git a/src/api/environment.cc b/src/api/environment.cc
index 0fb750c5ab..02e2a8f54b 100644
--- a/src/api/environment.cc
+++ b/src/api/environment.cc
@@ -386,6 +386,14 @@ void FreeEnvironment(Environment* env) {
   delete env;
 }
 
+NODE_EXTERN void Mount(Environment* env, const char *src, const char *dst, const int mode) {
+  env->VFS()->Mount(src, dst, mode);
+}
+
+NODE_EXTERN void Chroot(Environment* env, const char *path) {
+  env->VFS()->Chroot(path);
+}
+
 NODE_EXTERN std::unique_ptr<InspectorParentHandle> GetInspectorParentHandle(
     Environment* env,
     ThreadId thread_id,
