diff --git a/src/node.h b/src/node.h
index 5b1404ff8e..ada755afdc 100644
--- a/src/node.h
+++ b/src/node.h
@@ -438,7 +438,9 @@ enum Flags : uint64_t {
   // $HOME/.node_modules and $NODE_PATH. This is used by standalone apps that
   // do not expect to have their behaviors changed because of globally
   // installed modules.
-  kNoGlobalSearchPaths = 1 << 7
+  kNoGlobalSearchPaths = 1 << 7,
+  // Set if Node.js is running for embedded environment
+  kOwnsEmbedded = 1 << 8
 };
 }  // namespace EnvironmentFlags
 
@@ -486,6 +488,9 @@ NODE_EXTERN v8::MaybeLocal<v8::Value> LoadEnvironment(
     const char* main_script_source_utf8);
 NODE_EXTERN void FreeEnvironment(Environment* env);
 
+NODE_EXTERN void Mount(Environment* env, const char *src, const char *dst, const int mode);
+NODE_EXTERN void Chroot(Environment* env, const char *path);
+
 // Set a callback that is called when process.exit() is called from JS,
 // overriding the default handler.
 // It receives the Environment* instance and the exit code as arguments.
