diff --git a/tools/v8_gypfiles/toolchain.gypi b/tools/v8_gypfiles/toolchain.gypi
index d10525af57..86ca03edbb 100644
--- a/tools/v8_gypfiles/toolchain.gypi
+++ b/tools/v8_gypfiles/toolchain.gypi
@@ -62,9 +62,9 @@
     'mips_use_msa%': 0,
 
     # Print to stdout on Android.
-    'v8_android_log_stdout%': 0,
+    'v8_android_log_stdout%': 1,
 
-    'v8_enable_backtrace%': 0,
+    'v8_enable_backtrace%': 1,
 
     # Enable profiling support. Only required on Windows.
     'v8_enable_prof%': 0,
