diff --git a/tools/v8_gypfiles/features.gypi b/tools/v8_gypfiles/features.gypi
index 365860ef94..a92fc89999 100644
--- a/tools/v8_gypfiles/features.gypi
+++ b/tools/v8_gypfiles/features.gypi
@@ -30,7 +30,7 @@
 {
   'variables': {
     'v8_target_arch%': '<(target_arch)',
-
+    'v8_target_os': '<(OS)',
     'v8_current_cpu%': '<(target_arch)',
 
     # Emulate GN variables
@@ -413,6 +413,12 @@
       ['v8_advanced_bigint_algorithms==1', {
         'defines': ['V8_ADVANCED_BIGINT_ALGORITHMS',],
       }],
+      ['v8_target_os == "android"', {
+        'defines': [
+          'V8_HAVE_TARGET_OS',
+          'V8_TARGET_OS_ANDROID',
+        ],
+      }],
     ],  # conditions
     'defines': [
       'V8_GYP_BUILD',
